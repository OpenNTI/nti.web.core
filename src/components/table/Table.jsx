import React from 'react';
import cx from 'classnames';

import { VariantGetter } from '../../system/utils/PropGetters';
import { Placeholder } from '../placeholder/Placeholder';

import Theme from './Table.theme.css';
import { Footer } from './Footer';
import { Header } from './Header';
import { Row } from './Row';
export { SimpleTableHeader } from './headers/Simple';

const getTableVariant = VariantGetter(['plain', 'ruled'], 'plain');

const ColumnToPlaceholder = column => {
	const PlaceholderColumn = () => {
		const Cmp = column.Placeholder ?? Placeholder.Text;

		return <Cmp />;
	};

	PlaceholderColumn.FooterComponent = column.FooterComponent;
	PlaceholderColumn.HeaderComponent = column.HeaderComponent;
	PlaceholderColumn.Name = column.Name;
	PlaceholderColumn.CSSClassName = column.CSSClassName;

	return PlaceholderColumn;
};

export const TablePlaceholder = ({ rows, columns, ...otherProps }) => (
	<Table
		items={Array.from({ length: rows })}
		columns={columns.map(ColumnToPlaceholder)}
		{...otherProps}
	/>
);

/**
 * A Generic Table component that takes two props: columns & items.
 *
 * To define a Column component, make its primary function render the cell in the
 * body of the table. Add static class properties called "HeaderComponent" and/or
 * "FooterComponent" if you need a header/footer. These static properties will be
 * used to flag whether or not to add the thead/tfoot.
 *
 * ```js
 * ExampleColumn.HeaderComponent = () => <div/>
 * function ExampleColumn ({item}) {
 *   return (
 *   	<div />
 *   );
 * }
 *
 * // or Class:
 *
 * class ExampleColumn extends React.Component {
 * 	static HeaderComponent = () => <div/>
 * 	static FooterComponent = () => <div/>
 * 	render () {
 * 		return (
 * 			<div />
 * 		);
 * 	}
 * }
 * ```
 *
 * @template T
 * @param {import('./Table').TableProps<T>} props
 * @returns {JSX.Element}
 */
export function Table({
	className,

	columns,
	items,

	rowClassName,
	onRowClick,

	...extraProps
}) {
	const [variant] = getTableVariant(extraProps);

	return (
		<table className={cx(className, Theme.table, Theme[variant])}>
			<Header
				{...{
					columns,
					...extraProps,
				}}
			/>
			<Body
				{...{
					columns,
					items,
					rowClassName,
					onRowClick,
					...extraProps,
				}}
			/>
			<Footer {...{ columns }} />
		</table>
	);
}

/**
 * @template T
 * @param {import('./Table').CommonTableProps<T> & import('./Table').TableBodyProps<T>} props
 * @returns {JSX.Element}
 */
function Body({ columns, items, rowClassName, onRowClick: onClick, ...props }) {
	return (
		<tbody>
			{
				/*allow for any iterable*/
				[...(items || [])].map((item, row) => (
					<Row
						key={row}
						className={rowClassName?.(item, row, items)}
						{...{ item, columns, onClick, ...props }}
					/>
				))
			}
		</tbody>
	);
}

/** @typedef {() => string} NameGetter */

/**
 * Apply static Type keys to the component so it is a basic column.
 *
 * @template T
 * @param {React.ComponentType<T>} CellComponent
 * @param {string | NameGetter} name
 * @param {string} sortOn
 * @returns {typeof CellComponent}
 */
Table.asBasicColumn = (CellComponent, name, sortOn = null) =>
	Object.assign(CellComponent, { Name: name, SortOn: sortOn });
