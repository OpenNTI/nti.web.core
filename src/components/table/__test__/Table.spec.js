/* eslint-env jest */
import React from 'react';
import { render } from '@testing-library/react';

import { Table } from '../Table';

describe('Table', () => {
	test('Basic', () => {
		const A = Table.asBasicColumn(({ item }) => item, 'A');
		const B = Table.asBasicColumn(({ item }) => item * 2, 'B');
		const c = render(<Table columns={[A, B]} items={[1, 2]} />);
		expect(c.asFragment()).toMatchSnapshot();
	});

	test('Basic With Sorting', () => {
		const A = Table.asBasicColumn(({ item }) => item, 'A', 'foo');
		const B = Table.asBasicColumn(({ item }) => item * 2, 'B', 'bar');

		const c = render(
			<Table
				columns={[A, B]}
				items={[1, 2]}
				sortOn="bar"
				sortDirection="ascending"
				onChangeSort={jest.fn()}
			/>
		);
		expect(c.asFragment()).toMatchSnapshot();
	});

	test.todo('Column can render its TD');
	test.todo('Extra props pass through');
	test.todo('CSSClassName');
	test.todo('rowClassName');
	test.todo('Custom Header');
	test.todo('Custom Footer');
});
