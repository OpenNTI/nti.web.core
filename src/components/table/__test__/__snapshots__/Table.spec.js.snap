// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Basic 1`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <thead>
      <tr>
        <th>
          <div>
            <span>
              A
            </span>
          </div>
        </th>
        <th>
          <div>
            <span>
              B
            </span>
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-name="A"
        >
          1
        </td>
        <td
          data-name="B"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          data-name="A"
        >
          2
        </td>
        <td
          data-name="B"
        >
          4
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table Basic With Sorting 1`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <thead>
      <tr>
        <th>
          <div
            class="sortable"
          >
            <span>
              A
            </span>
          </div>
        </th>
        <th>
          <div
            class="sortable sorted asc"
          >
            <span>
              B
            </span>
            <i
              class="icon-chevron-down"
            />
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-name="A"
        >
          1
        </td>
        <td
          data-name="B"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          data-name="A"
        >
          2
        </td>
        <td
          data-name="B"
        >
          4
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table Basic With Sorting 2`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <thead>
      <tr>
        <th>
          <div
            class="sortable enabled"
          >
            <span>
              A
            </span>
          </div>
        </th>
        <th>
          <div
            class="sortable sorted asc enabled"
          >
            <span>
              B
            </span>
            <i
              class="icon-chevron-down"
            />
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-name="A"
        >
          1
        </td>
        <td
          data-name="B"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          data-name="A"
        >
          2
        </td>
        <td
          data-name="B"
        >
          4
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table Basic With Sorting 3`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <thead>
      <tr>
        <th>
          <div
            class="sortable enabled"
          >
            <span>
              A
            </span>
          </div>
        </th>
        <th>
          <div
            class="sortable sorted desc enabled"
          >
            <span>
              B
            </span>
            <i
              class="icon-chevron-up"
            />
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-name="A"
        >
          1
        </td>
        <td
          data-name="B"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          data-name="A"
        >
          2
        </td>
        <td
          data-name="B"
        >
          4
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table Column can render its TD 1`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <tbody>
      <tr>
        <td
          data-custom="td"
        >
          foo
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table Custom Footer 1`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <tbody>
      <tr>
        <td>
          foo
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <th>
          <div
            class="custom"
            data-footer="custom"
          >
            My Footer
          </div>
        </th>
      </tr>
    </tfoot>
  </table>
</DocumentFragment>
`;

exports[`Table Custom Header 1`] = `
<DocumentFragment>
  <table
    class="table plain"
  >
    <thead>
      <tr>
        <th>
          <div
            class="custom"
            data-header="custom"
          >
            My Header
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          foo
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
